src/ui/localization/translator.ts

```diff
@@ -0,0 +1,246 @@
+ * This class allows you to override the default translations used in the purchase flow.
+ * The main level keys are the locale codes and the values are objects with the same keys as the default translations.
+ * @public
+ * @example
+ * This example will override the default translation for the email step title in the English locale.
+ * ```typescript
+ * const customTranslations = {
+ *  en: {
+ *    "state_needs_auth_info.email_step_title": "Billing email",
+ *   }
+ * }
+ * ```
+ */
+export type CustomTranslations = Record<string, Record<string, string>>;
+
+/**
+ * Translation variables to be used in the translation.
+ * This class allows you to pass variables to the translate method.
+ * @public
+ * @example Given a label with id `periods.monthPlural` and value `{{amount}} months`. This example will replace the variable `{{amount}}` with the value `10`.
+ * ```typescript
+ * translator.translate('periods.monthPlural', { amount: 10 });
+ * // Output: 10 months
+ * ```
+ */
+export type TranslationVariables = Record<
+  string,
+  string | number | undefined | null
+>;
+
+export interface TranslatePeriodOptions {
+  noWhitespace?: boolean;
+  short?: boolean;
+}
+
+const defaultTranslatePeriodOptions: TranslatePeriodOptions = {
+  noWhitespace: false,
+  short: false,
+};
+
+export class Translator {
+  public readonly locales: Record<string, LocaleTranslations> = {};
+
+  public static fallback() {
+    return new Translator();
+  }
+
+  public constructor(
+    customTranslations: CustomTranslations = {},
+    public readonly selectedLocale: string = "en",
+    public readonly defaultLocale: string = "en",
+  ) {
+    this.locales = {
+      en: new LocaleTranslations(en, "en"),
+      es: new LocaleTranslations(es, "es"),
+      it: new LocaleTranslations(it, "it"),
+    };
+    if (customTranslations) {
+      this.override(customTranslations);
+    }
+  }
+
+  public override(customTranslations: CustomTranslations) {
+    Object.entries(customTranslations).forEach(([locale, translations]) => {
+      this.locales[locale] = new LocaleTranslations(
+        {
+          ...(this.locales[locale].labels || {}),
+          ...translations,
+        },
+        this.simplifyLocaleString(locale),
+      );
+    });
+  }
+
+  public formatPrice(priceInMicros: number, currency: string): string {
+    const price = priceInMicros / 1000000;
+    const formatter = new Intl.NumberFormat(this.locale, {
+      style: "currency",
+      currency,
+    });
+
+    return formatter.format(price);
+  }
+
+  get locale(): string {
+    return (
+      this.getLocaleInstance(this.selectedLocale)?.localeKey ||
+      this.simplifyLocaleString(this.selectedLocale)
+    );
+  }
+
+  get fallbackLocale(): string {
+    return (
+      this.getLocaleInstance(this.defaultLocale)?.localeKey ||
+      this.simplifyLocaleString(this.defaultLocale)
+    );
+  }
+
+  private simplifyLocaleString(locale: string): string {
```

Comment:
Nitpick but how about renaming to `getLanguageCodeString` or something like that? I believe it would be more precise that way.

Reply:
Will do!

Reply:
Renamed